src/rulesets/Base/broadcast.jl

```diff
@@ -0,0 +1,369 @@
+using Base.Broadcast: Broadcast, broadcasted, Broadcasted
+const RCR = RuleConfig{>:HasReverseMode}
+
+function rrule(::typeof(copy), bc::Broadcasted)
+    uncopy(Δ) = (NoTangent(), Δ)
+    return copy(bc), uncopy
+end
+
+# Skip AD'ing through the axis computation
+function rrule(::typeof(Broadcast.instantiate), bc::Broadcasted)
+    uninstantiate(Δ) = (NoTangent(), Δ)
+    return Broadcast.instantiate(bc), uninstantiate
+end
+
+_print(args...) = nothing # println(join(args, " ")) # 
+
+#####
+##### Split broadcasting
+#####
+
+# For `z = g.(f.(xs))`, this finds `y = f.(x)` eagerly because the rules for either `f` or `g` may need it,
+# and we don't know whether re-computing `y` is cheap. 
+# (We could check `f` first like `sum(f, x)` does, but checking whether `g` needs `y` is tricky.)
+
+function rrule(cfg::RCR, ::typeof(broadcasted), f::F, args::Vararg{Any,N}) where {F,N}
```

Comment:
This rule is applied before Zygote's own rule. To avoid that, it could be changed to this:
```suggestion
function rrule(cfg::RCR, ::typeof(broadcasted), ::BroadcastStyle, f::F, args...) where {F}
```
This has two problems. 

The first is that Yota no longer sees the rule at all. Trying to make it a specific rule which inserts the style is tricky, as this creates method ambiguities:
```
function ChainRulesCore.rrule(cfg::YotaRuleConfig, ::typeof(Broadcast.broadcasted), f::F, args::Vararg{Any,N}) where {F,N}
    bcs = Broadcast.BroadcastStyle(typeof(Broadcast.broadcasted(f, args...)))
    y, back = ChainRulesCore.rrule(cfg, Broadcast.broadcasted, bcs, f, args...)  # this one is ambiguous
...
```
@dfdx is there a better way?

The second problem is that this (slightly later) method has its arguments wrapped by `broadcastable`, which means many things need `Ref`:
```julia
julia> using Base.Broadcast: broadcasted, DefaultArrayStyle, broadcastable

julia> copy(broadcasted(|>, [1,2,3], inv))
3-element Vector{Float64}:
 1.0
 0.5
 0.3333333333333333

julia> copy(broadcasted(DefaultArrayStyle{1}(), |>, [1,2,3], Ref(inv)))  # error without Ref
3-element Vector{Float64}:
 1.0
 0.5
 0.3333333333333333

julia> broadcastable(inv)
Base.RefValue{typeof(inv)}(inv)
```
Which at the moment gives test errors like `Tangent{Base.RefValue{typeof(sin)}}(x = NoTangent(),) isa NoTangent`, lots of messy Tangents.

Reply:
Perhaps better is for Zygote to be told not to use these rules, via `@opt_out`. Old versions of Zygote won't have that, and would perhaps need to get an upper version bound on ChainRules.

Reply:
I think we should be able to do a configured opt-out for ZygoteConfig

Reply:
> Which at the moment gives test errors like Tangent{Base.RefValue{typeof(sin)}}(x = NoTangent(),) isa NoTangent, lots of messy Tangents.

Can we add a `simplify` call at the end which converts `all(isa(NoTangent), x::Tangent)` into `NoTangent()` ?

Reply:
The opt-out is https://github.com/FluxML/Zygote.jl/pull/1263 BTW.

Reply:
> The first is that Yota no longer sees the rule at all. Trying to make it a specific rule which inserts the style is tricky, as this creates method ambiguities

If needed, I can hack broadcast in Yota to always have the `BroadcastStyle` argument - just let me know once the decision is made.

Yet, I'm not sure it will be a good approach for other similar ADs. If I understand it correctly:

1. Both - Yota and Zygote - parse the lowered code of a function to extract all calls.
2. They both encounter a call to `broadcasted(f, args...)`.
3. They rewrite it to `rrule(cfg, broadcasted, f, args...)`.

`BroadcastStyle` is never encountered or added explicitly, so neither Zygote nor Yota ever touch `rrule()` with it. I'd expect other ADs working with the lowered code to behave the same way and hit the same issue.

Reply:
Ok. Yes I think it would be simplest never to have the style in the rules. Sorry I guess thinking through the Zygote problem out loud... but opting out there seems simplest.

The simpler rules here should work quite well on CuArray broadcasts. If a special path is needed for them sometimes, it can dispatch later somehow, not on this `rrule` signature.

Reply:
Re-opening since opt-out seems more delicate than anticipated. 

Using the macro in the obvious way to opt out of the generic rule (by matching its signature) produces ambiguities with every more specific rule. Instead, it might work to overload only `no_rrule` as here:

https://github.com/FluxML/Zygote.jl/pull/1263/commits/a77f6123151dc9ea9f4155d3583bbdea198bc2de

Not sure whether that's legal. 

Also not sure whether that's enough. When last I tried, Zygote's tests called quite a few of the rules here, including the generic one. I think that possibly `rrule_via_ad` takes a shortcut which does not check `no_rrule`.

Reply:
BTW, why does `@opt_out` define `no_rrule` to return `nothing`, which it already returns? That means you have to check the method table, but it seems you could equally well return `true` or `NoTangent` or something, which the caller can easily test. Can we fix that?

Reply:
I am going to need to refresh my memory on how and why `@opt_out` works the way it does.  :sweat_smile: 

> I think that possibly rrule_via_ad takes a shortcut which does not check no_rrule.

There is indeed a shortcut that doesn't check `no_rule`
When it goes through the path of Zygote using Zygote because these is no rule then it hits the normal zygote check.
But there is a fast path that avoids calling back in to zygote if there is a rule right there.
https://github.com/FluxML/Zygote.jl/blob/5c80f55b24d4060f5550b25201b4b288703d34b9/src/compiler/chainrules.jl#L243-L245

But that path does check if `rrule` returns `nothing`.
Which it should as, the `@opt_out` defines a more specific `rrule` that returns `nothing` for this case.
So there should be no need to check `no_rrule`.
But that doesn't hold if you bypass `@opt_out`, which is why the docs are specific about not being able to bypass `@opt_out` as you can't know if user is using the method table way, or the directly checking `rrule` way.

> BTW, why does `@opt_out` define `no_rrule` to return nothing, which it already returns?

It doesn't return anything because it is just using the method table for storage.
But the user shouldn't need to check `no_rrule` because checking that `rrule` returns `nothing` rather than a result should be enough.

> Using the macro in the obvious way to opt out of the generic rule (by matching its signature) produces ambiguities with every more specific rule

Aren't these real ambiguities?
Can you show some examples?

Reply:
This PR defines one generic, and many specific rules:
```julia
rrule(::RuleConfig{>:HasReverseMode}, ::typeof(broadcasted), f::F, args::Vararg{Any,N})  # generic
rrule(::RuleConfig{>:HasReverseMode}, ::typeof(broadcasted), ::typeof(+), xs::Union{Number, Broadcasted, AbstractArray{<:Number}, Tuple{Vararg{Number, N}}}...) # specific
```
Attempting to opt out of the generic rule for Zygote, the macro produces this:
```julia
rrule(::Zygote.ZygoteRuleConfig, ::typeof(broadcasted), f::F, args::Vararg{Any, N}) where {F, N}
```
which is more specific on the 1st but less specific on the 3rd argument, sometimes.

The hope was to opt out of only the generic rule, since the rules for `+` etc. are similar enough not to matter. But the generic one is quite different, and e.g. skips Zygote's present CuArray handling... I'd rather not touch that yet.

Such an opt-out could be done if you are allowed to define only `no_rrule` (and the caller expected to check that). Note that the only caller here is Zygote, but still. 

Maybe the Zygote PR does have to opt out of every rule here, I think that would remove ambiguity. The tricky question is what happens to 3rd party packages defining broadcasting rules. NNlib has
```julia
rrule(::typeof(broadcasted), ::typeof(relu), x::Numeric)
```
which won't in fact cause an ambiguity. But it also won't ever be called, with this PR, since the `::RuleConfig` method will match first. It could be given a `::RuleConfig` argument (as the specific rules are in this PR) but then the ambiguity will return.

Reply:
[7e7d105](https://github.com/JuliaDiff/ChainRules.jl/pull/644/commits/7e7d1054c557da711a784d0c5d92f6b6fc8eb475) changes over to use BroadcastStyle. 

I now think this is the right thing to do. Besides Zygote, the generic rule needs `RuleConfig{>:HasReverseMode}` while the specific rules do not. Naiively that would mean the specific rules never get called, so previously I gave them a RuleConfig argument only for this reason. But that's a bit weird, and e.g. NNlib's rules don't, https://github.com/FluxML/NNlib.jl/blob/master/src/activations.jl#L875-L880 .

Now, instead, when no specific rule matches, AD will continue & hit the rule with both RuleConfig and Broadcasted. (For some reason this didn't work with Yota, earlier, but surely fixable.)

Reply:
Zygote does appear to occasionally call some of these rules, in its tests. Perhaps because of `rrule_via_ad`? But they aren't so different, and I think tests pass.